### PR TITLE
Allow partial shape information in `SpecifyShape` `Op`

### PR DESCRIPTION
**Pull request by ricardoV94**
Shape values of `None` can be used for dimensions that are not fixed

This seemed necessary to replace the more limited use of Rebroadcast in `convert_variable`:
https://github.com/aesara-devs/aesara/blob/c34f37f140a4e3c2006107b7a4dedc26064bb249/aesara/tensor/type.py#L326-L329

The ultimate goal, would be to allow us to work more gradually towards #732, such as in #821

***
Misc ideas:
* Leave original `SpecifyShape` intact and create new `SpecifyPartialShape` (some esoteric optimizations that used to work no longer do, see disabled subtensor tests)

### CODE REVIEW

**Comment by ricardoV94** [comment]
It seems `infer_shape` and `type.shape` need to be more closely integrated... but I don't have a good mental picture of what `infer_shape` and `.shape` are doing to be able to say more.

**Review by brandonwillard — CHANGES_REQUESTED** [review]
This is a change mentioned in https://github.com/aesara-devs/aesara/issues/748, where `None` is the proposed indicator of non-static shape information&mdash;as it is in general.  

These changes definitely should use `None` instead of `-1`, if only because we don't want two different encodings for the exact same concept, but, also, because we want to support expressions like `specify_shape(new_x, old_x.type.shape)`.  That's exactly the kind of expression we might want to use to replace `Rebroadcast` in `TensorType.convert_variable` and close https://github.com/aesara-devs/aesara/issues/748.

File: aesara/tensor/shape.py [hunk 1]
```diff
@@ -411,12 +430,10 @@ def make_node(self, x, shape):
...
-        shape = at.as_tensor_variable(shape, ndim=1)
-
-        if isinstance(shape, Constant):
-            shape = tuple(shape.data)
-        else:
-            shape = tuple(at.as_tensor_variable(s, ndim=0) for s in shape)
+        shape = tuple(
+            at.as_tensor_variable(s, ndim=0)
+            for s in at.as_tensor_variable(shape, ndim=1)
+        )
```
Comment (brandonwillard) [inline aesara/tensor/shape.py:436]: This will introduce unnecessary `Subtensor` operations when `shape` is a constant, and an increased need for constant folding.

File: aesara/tensor/shape.py [hunk 2]
```diff
@@ -426,13 +443,18 @@ def make_node(self, x, shape):
...
 
-        if isinstance(x.type, TensorType) and all(isinstance(s, Number) for s in shape):
-            out_var = TensorType(x.type.dtype, shape)()
+        if isinstance(x.type, TensorType):
+            mixed_shape = tuple(
+                xts if xts is not None else s for xts, s in zip(x.type.shape, shape)
+            )
+            folded_shape = constant_fold_shape(mixed_shape)
+            folded_valid_shape = tuple(s if s != -1 else None for s in folded_shape)
+            out_var = TensorType(x.type.dtype, folded_valid_shape)()
```
Comment (brandonwillard) [inline aesara/tensor/shape.py:452]: The comment above mentions one way in which the need for constant folding can be reduced, but, more importantly, we really shouldn't need constant folding in order to find the static/non-static shape values.

Instead, if we use the established `None` encoding, then we can use a direct `NoneConst.equals` comparison with other symbolic terms (like the elements of `shape`) and it will work without constant folding.

**Review by brandonwillard — CHANGES_REQUESTED** [review]

**Comment by ricardoV94** [comment]
> These changes definitely should use `None` instead of `-1`, if only because we don't want two different encodings for the exact same concept

This wouldn't work because `NoneConst` is not allowed in vectors AFAICT

```python
import aesara.tensor as at
shape = at.as_tensor_variable([at.NoneConst, 1])
# TypeError: Unsupported dtype for TensorType: object
```

**Comment by brandonwillard** [comment]
> > These changes definitely should use `None` instead of `-1`, if only because we don't want two different encodings for the exact same concept
> 
> This wouldn't work because `NoneConst` is not allowed in vectors AFAICT
> 
> ```python
> import aesara.tensor as at
> shape = at.as_tensor_variable([at.NoneConst, 1])
> # TypeError: Unsupported dtype for TensorType: object
> ```

You've just identified a real long-standing issue with the design of `SpecifyShape`: it purports to work with variable length vectors, but actually doesn't/shouldn't.  In other words, `SpecifyShape`'s shape inputs are necessarily fixed-length, the only problem is that its design doesn't match that fact, and, as a result, some things are made unnecessarily difficult to implement/change.

File: tests/tensor/test_subtensor_opt.py [hunk 3]
```diff
@@ -127,6 +127,7 @@ def test_local_replace_AdvancedSubtensor(indices, is_none):
         assert np.array_equal(res_val, exp_res_val)
 
 
+@pytest.mark.xfail(reason="Temporarily disabled")
```
Comment (ricardoV94) [inline tests/tensor/test_subtensor_opt.py:130]: Couldn't quite pinpoint the problem yet. The imported_by is not so informative, as it mentions a MergeOptimizer with +30 rewrites. I suspect it has something to do with the use of `ShapeFeature.same_shape` that needs some tweaking. Help appreciated

File: tests/link/test_numba.py [hunk 4]
```diff
@@ -878,6 +878,7 @@ def test_Reshape_scalar():
     )
 
 
+@pytest.mark.xfail(reason="Fails with Nones")
```
Comment (ricardoV94) [inline tests/link/test_numba.py:881]: Couldn't figure out how to get Numba to work with Nones

File: aesara/tensor/shape.py [hunk 5]
```diff
@@ -397,169 +377,197 @@ class SpecifyShape(COp):
     -----
     Maybe in the future we will never do the assert!
 
-    We currently don't support specifying partial shape information.
-
-    TODO : test this op with sparse. Do C code for them too.
-
+    TODO: Implement rewrite to remove useless SpecifyShape when:
+        1. Input shape is already fully known
+        2. All unknown input shapes are associated with NoneConst
```
Comment (ricardoV94) [inline aesara/tensor/shape.py:382]: Maybe this does not quite make sense... depends on whether Aesara complains if an intermediate variable of fixed shape is given a value with the wrong shape

Comment (ricardoV94) [reply aesara/tensor/shape.py:382]: We can also narrow down the scope of such useless rewrite, to the cases where the not-none specified shape are constants, because then we can eagerly check if they match with the input type shape. This is also the most common use of SpecifyShape

File: tests/link/test_numba.py (hunk 4 above)
Comment (ricardoV94) [reply tests/link/test_numba.py:881]: It's a compilation issue not just a `compare_numba_and_py`.

```python
import aesara
import aesara.tensor as at
import numpy as np
x = at.matrix("x")
y = at.specify_shape(x, (1, None))
f = aesara.function([x], y, mode="NUMBA")
f(np.zeros((1, 2)))
```

<details>
 <summary>Traceback</summary>

```
Traceback (most recent call last):
  File "/home/ricardo/Documents/Projects/aesara/aesara/link/utils.py", line 202, in streamline_default_f
    thunk()
  File "/home/ricardo/Documents/Projects/aesara/aesara/link/basic.py", line 664, in thunk
    outputs = fgraph_jit(*[x[0] for x in thunk_inputs])
  File "/home/ricardo/Documents/Projects/aesara/venv/lib/python3.8/site-packages/numba/core/dispatcher.py", line 482, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/home/ricardo/Documents/Projects/aesara/venv/lib/python3.8/site-packages/numba/core/dispatcher.py", line 423, in error_rewrite
    raise e.with_traceback(None)
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function(<class 'zip'>) found for signature:
 
 >>> zip(UniTuple(int64 x 2), StarArgTuple(readonly array(int8, 0d, C), none))
 
There are 2 candidate implementations:
  - Of which 2 did not match due to:
  Overload of function 'zip': File: numba/core/typing/builtins.py: Line 1019.
    With argument(s): '(UniTuple(int64 x 2), StarArgTuple(readonly array(int8, 0d, C), none))':
   No match.
During: resolving callee type: Function(<class 'zip'>)
During: typing of call at /home/ricardo/Documents/Projects/aesara/aesara/link/numba/dispatch/basic.py (589)
File "aesara/link/numba/dispatch/basic.py", line 589:
    def specifyshape(x, *shape):
        for xs, s in zip(x.shape, shape):
        ^
During: resolving callee type: type(CPUDispatcher(<function numba_funcify_SpecifyShape.<locals>.specifyshape at 0x7f11594b48b0>))
During: typing of call at /tmp/tmphikq3lce (3)
During: resolving callee type: type(CPUDispatcher(<function numba_funcify_SpecifyShape.<locals>.specifyshape at 0x7f11594b48b0>))
During: typing of call at /tmp/tmphikq3lce (3)
File "../../../../../tmp/tmphikq3lce", line 3:
def numba_funcified_fgraph(x):
    auto_59 = specifyshape(x, auto_55, NoneConst)
    ^
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/ricardo/Documents/Projects/aesara/venv/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3441, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-4-462a1a0bd99b>", line 7, in <module>
    f(np.zeros((1, 2)))
  File "/home/ricardo/Documents/Projects/aesara/aesara/compile/function/types.py", line 964, in __call__
    self.fn()
  File "/home/ricardo/Documents/Projects/aesara/aesara/link/utils.py", line 206, in streamline_default_f
    raise_with_op(fgraph, node, thunk)
  File "/home/ricardo/Documents/Projects/aesara/aesara/link/utils.py", line 538, in raise_with_op
    raise exc_value.with_traceback(exc_trace)
  File "/home/ricardo/Documents/Projects/aesara/aesara/link/utils.py", line 202, in streamline_default_f
    thunk()
  File "/home/ricardo/Documents/Projects/aesara/aesara/link/basic.py", line 664, in thunk
    outputs = fgraph_jit(*[x[0] for x in thunk_inputs])
  File "/home/ricardo/Documents/Projects/aesara/venv/lib/python3.8/site-packages/numba/core/dispatcher.py", line 482, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/home/ricardo/Documents/Projects/aesara/venv/lib/python3.8/site-packages/numba/core/dispatcher.py", line 423, in error_rewrite
    raise e.with_traceback(None)
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function(<class 'zip'>) found for signature:
 
 >>> zip(UniTuple(int64 x 2), StarArgTuple(readonly array(int8, 0d, C), none))
 
There are 2 candidate implementations:
  - Of which 2 did not match due to:
  Overload of function 'zip': File: numba/core/typing/builtins.py: Line 1019.
    With argument(s): '(UniTuple(int64 x 2), StarArgTuple(readonly array(int8, 0d, C), none))':
   No match.
During: resolving callee type: Function(<class 'zip'>)
During: typing of call at /home/ricardo/Documents/Projects/aesara/aesara/link/numba/dispatch/basic.py (589)
File "aesara/link/numba/dispatch/basic.py", line 589:
    def specifyshape(x, *shape):
        for xs, s in zip(x.shape, shape):
        ^
During: resolving callee type: type(CPUDispatcher(<function numba_funcify_SpecifyShape.<locals>.specifyshape at 0x7f11594b48b0>))
During: typing of call at /tmp/tmphikq3lce (3)
During: resolving callee type: type(CPUDispatcher(<function numba_funcify_SpecifyShape.<locals>.specifyshape at 0x7f11594b48b0>))
During: typing of call at /tmp/tmphikq3lce (3)
File "../../../../../tmp/tmphikq3lce", line 3:
def numba_funcified_fgraph(x):
    auto_59 = specifyshape(x, auto_55, NoneConst)
    ^
Apply node that caused the error: DeepCopyOp(SpecifyShape.0)
Toposort index: 1
Inputs types: [TensorType(float64, (1, None))]
Inputs shapes: [(1, 2)]
Inputs strides: [(16, 8)]
Inputs values: [array([[0., 0.]])]
Outputs clients: [['output']]
HINT: Re-running with most Aesara optimizations disabled could provide a back-trace showing when this node was created. This can be done by setting the Aesara flag 'optimizer=fast_compile'. If that does not work, Aesara optimizations can be disabled with 'optimizer=None'.
HINT: Use the Aesara flag `exception_verbosity=high` for a debug print-out and storage map footprint of this Apply node.
```
</details>

Comment (ricardoV94) [reply tests/link/test_numba.py:881]: I tried a bunch of things, include creating and index of non-None inputs outside and inside the function, but nothing worked for me. Just different errors.

File: aesara/tensor/shape.py [hunk 6]
```diff
@@ -396,136 +376,157 @@ class SpecifyShape(COp):
...
+                    try:
+                        type_s = get_scalar_constant_value(s)
+                        if type_s is not None:
+                            type_shape[i] = int(type_s)
+                    except NotScalarConstantError:
+                        pass
+
+            out_var = x.type.clone(shape=type_shape)()
         else:
             out_var = x.type()
```
[the remark this answers is not in the document]
Comment (ricardoV94) [reply aesara/tensor/shape.py:425]: Right, but SpecifyShape does not work for those

[the remark this answers is not in the document]
Comment (brandonwillard) [reply aesara/tensor/shape.py:425]: Yeah, and the `as_tensor_variable` above should also only ever return `TensorType` results.  It looks like the acceptable types should be clarified upfront (e.g. an explicit exception for non-`TensorType`s).

[the remark this answers is not in the document]
Comment (ricardoV94) [reply aesara/tensor/shape.py:425]: Is it enough to try to convert everything to `as_tensor_variable`? I assume that will raise for non `TensorType`s

File: aesara/tensor/shape.py [hunk 7]
```diff
@@ -396,136 +376,153 @@ class SpecifyShape(COp):
...
         (gz,) = grads
         # Should I set an SpecifyShape on gz? I think so
         # But I don't do it now as we need to make an optimization
         # to remove that op from the graph to don't block other optimization
         # Should I do an optimizer that will remove the SpecifyShape?
         # I think Yes
-        return [gz, aesara.gradient.DisconnectedType()()]
-        return [specify_shape(gz, s), aesara.gradient.DisconnectedType()()]
+        # return [specify_shape(gz, s)] + [aesara.gradient.DisconnectedType()()] * len(shape)
+        return [gz] + [aesara.gradient.DisconnectedType()()] * len(shape)
```
Comment (ricardoV94) [inline aesara/tensor/shape.py:467]: Check: Do I need a unique DisconnectedType per input or is the multiplication fine

File: aesara/tensor/shape.py [hunk 8]
```diff
@@ -537,29 +534,31 @@ def specify_shape(
         int, List[Union[int, Variable]], Tuple[Union[int, Variable]], Variable
     ],
 ):
-    """Specify a fixed shape for a `Variable`."""
+    """Specify a fixed shape for a `Variable`.
```
Comment (ricardoV94) [inline aesara/tensor/shape.py:537]: Do we need this helper function? Can't the code go into make_node?

File: aesara/tensor/shape.py (hunk 7 above)
Comment (brandonwillard) [reply aesara/tensor/shape.py:467]: I'm not entirely sure how the `Variable` instances of these are used in every scenario, but I do see that distinct ones are created for each distinct term to which they correspond, so it would be safer to create unique ones.

File: aesara/tensor/shape.py (hunk 8 above)
Comment (brandonwillard) [reply aesara/tensor/shape.py:537]: Not if it would make the `Op.make_node` and `Apply.inputs` inconsistent.  It looks like this `specify_shape` helper/constructor function has an `(x, shape)` signature and not `(x, *shape)`, so that might be one reason not to get rid of it.

File: tests/link/test_numba.py (hunk 4 above)
Comment (brandonwillard) [reply tests/link/test_numba.py:881]: Is this still an issue?

File: aesara/tensor/shape.py (hunk 8 above)
Comment (ricardoV94) [reply aesara/tensor/shape.py:537]: My main concern was to make sure that `SpecifyShape` `Op` would work with it's own inputs so that you can call `node.owner.op(*node.owner.inputs)`. I guess once we have that, it doesn't matter we have this helper logic separated.

File: tests/link/test_numba.py (hunk 4 above)
Comment (ricardoV94) [reply tests/link/test_numba.py:886]: Yes. Kaustubh wrote this implementation, but it seems pretty involved:

```python
@numba_funcify.register(SpecifyShape)
def numba_funcify_SpecifyShape(op, node, **kwargs):
    
    num_shape_args = len(node.inputs) - 1
    input_names = ["shape_" + str(i) for i in range(num_shape_args)]
    shape_indices = [f"x.shape[{i}]" for i in range(num_shape_args)]

    func_body = ""
    for _input_name, _shape_index in zip(input_names, shape_indices):
        func_body = func_body + f"""
    if {_input_name} is not None:
        assert {_shape_index} == {_input_name}
"""

    func_template = f"""
def specifyshape(x, {create_arg_string(input_names)}):
    {func_body}
    return x
"""
    specifyshape = compile_function_src(
        func_template, "specifyshape", globals()
    )
    return numba_njit(specifyshape)
```

Comment (ricardoV94) [reply tests/link/test_numba.py:886]: I suggested instead doing o numba-specific rewrite where we replace `None` by `-1` just before compilation, which would get rid of the issues with mixed arg types that Numba seems to reject. I don't have a strong preference for either option. What do you think?

File: aesara/tensor/shape.py (hunk 8 above)
Comment (brandonwillard) [reply aesara/tensor/shape.py:537]: > My main concern was to make sure that `SpecifyShape` `Op` would work with it's own inputs so that you can call `node.owner.op(*node.owner.inputs)`.

Yes, that is by far the most important thing.

More generally, it would be great if we could confirm this automatically somehow (e.g. via type checking, but that doesn't seem likely).

**Comment by ricardoV94** [comment]
Numba implemetation is working, only issues left are the two failing rewrite tests

File: aesara/tensor/shape.py [hunk 9]
```diff
@@ -396,136 +376,157 @@ class SpecifyShape(COp):
...
             try:
-                s = at.get_scalar_constant_value(node.inputs[1][dim])
-                s = at.as_tensor_variable(s)
-                new_shape.append(s)
+                s = at.get_scalar_constant_value(s)
+                # We assume that `None` shapes are always retrieved by
+                # `get_scalar_constant_value`, and only in that case do we default to
+                # the shape of the input variable
+                if s is None:
+                    s = xshape[dim]
```
Comment (ricardoV94) [inline aesara/tensor/shape.py:451]: The failing rewrites were due to me being perhaps over-careful here and defaulting to using the shape of the input variable `xshape` whenever `at.get_scalar_constant_value` failed to return a Constant. I was doing this because I feared a hidden `None` specified shape could be lurking somewhere in the graph of the shape. 

However, since most Op's cannot build on top of `NoneConst` inputs I have now switched to assuming that any error also confirms that the specified shape is not `None` and can be used for shape inference in this `Op`. The rewrites now work, but let's see if anything else fails

Comment (ricardoV94) [reply aesara/tensor/shape.py:451]: Seems to work

File: aesara/tensor/subtensor_opt.py [hunk 10]
```diff
@@ -1663,6 +1664,11 @@ def local_subtensor_SpecifyShape_lift(fgraph, node):
     if new_obj_arg.ndim == 0:
         new_shape_arg = as_tensor([], dtype=np.int64)
     else:
-        new_shape_arg = as_tensor(shape_arg[indices], dtype=np.int64, ndim=1)
+        # We cannot stack the shape arguments if there are None in it.
```
[the remark this answers is not in the document]
Comment (ricardoV94) [reply aesara/tensor/subtensor_opt.py:1667]: Actually this rewrite is wrong. `x[i].shape != x.shape[i]`

Furthermore, this rewrite is only applied when there is a single non-slice index, which means that regardless of the index, it should always return `specify_shape(new_obj_arg, *shape_arg[1:])`

[the remark this answers is not in the document]
Comment (ricardoV94) [reply aesara/tensor/subtensor_opt.py:1667]: This test does not show this: https://github.com/aesara-devs/aesara/blob/0f1f5bebfb2a058540d1a54b7026b962c13f873a/tests/tensor/test_subtensor_opt.py#L2102-L2121

Adding this parametrization, makes it obvious:
```python
        (
            matrix(),
            (iscalar(), iscalar()),
            (0,),
            np.array([[1, 2, 3], [3, 4, 3]], dtype=config.floatX),
            np.array([2, 3], dtype=np.int64),
        ),
```

[the remark this answers is not in the document]
Comment (ricardoV94) [reply aesara/tensor/subtensor_opt.py:1667]: Fixed now

File: aesara/tensor/subtensor_opt.py [hunk 11]
```diff
@@ -1646,23 +1646,18 @@ def local_subtensor_SpecifyShape_lift(fgraph, node):
...
+    shape_arg = specify_shape_node.owner.inputs[1:]
 
     indices = get_idx_list(node.inputs, node.op.idx_list)
 
-    if len(indices) > 1 or len(indices) == 0:
-        return False
-
-    if isinstance(indices[0], slice) or isinstance(
-        getattr(indices[0], "type", None), SliceType
+    if any(
```
Comment (ricardoV94) [inline aesara/tensor/subtensor_opt.py:1653]: We could also handle mixed `[:]` slices easily, where we just have to keep the respective dimensions shape, but that's a bit beyond the  scope of this PR. 

Something like 
```python
# before:
specify_shape(x, (s1, s2))[:, 0 ]

#after:
specify_shape(x[:, 0], (s1,))
```

File: aesara/tensor/subtensor_opt.py [hunk 12]
```diff
@@ -1646,23 +1646,18 @@ def local_subtensor_SpecifyShape_lift(fgraph, node):
...
+        isinstance(index, slice) or isinstance(getattr(index, "type", None), SliceType)
+        for index in indices
```
Comment (brandonwillard) [inline aesara/tensor/subtensor_opt.py:1655]: Just as I mentioned above, whenever we encounter a `None` we can use `aesara.tensor.shape.shape_i`.

File: tests/tensor/test_shape.py [hunk 13]
```diff
@@ -441,19 +465,39 @@ def test_bad_shape(self):
...
+        with pytest.raises(AssertionError, match="SpecifyShape:.*"):
+            f(xval, 3)
+
     def test_infer_shape(self):
         rng = np.random.default_rng(3453)
         adtens4 = dtensor4()
         aivec = TensorVariable(TensorType("int64", (4,)))
         aivec_val = [3, 4, 2, 5]
-        adtens4_val = rng.random(aivec_val)
+        adtens4_val = rng.random((3, 4, 2, 5))
```
Comment (brandonwillard) [inline tests/tensor/test_shape.py:480]: Why this change?

**Review by brandonwillard — CHANGES_REQUESTED** [review]

File: aesara/tensor/subtensor_opt.py (hunk 12 above)
Comment (ricardoV94) [reply aesara/tensor/subtensor_opt.py:1655]: This require no longer relies on specifyshape input types, it doesn't matter if it's a `None` or not.

File: tests/tensor/test_shape.py (hunk 13 above)
Comment (ricardoV94) [reply tests/tensor/test_shape.py:480]: Looks like a mistaken leftover

File: aesara/tensor/basic_opt.py [hunk 14]
```diff
@@ -3521,7 +3523,12 @@ def local_Shape_of_SpecifyShape(fgraph, node):
     if not isinstance(getattr(specified_shape.owner, "op", None), SpecifyShape):
         return False
 
-    return [specified_shape.owner.inputs[1].astype(np.int64)]
+    shape = specified_shape.owner.inputs[1:]
+
+    if any(NoneConst.equals(s) for s in shape):
+        return False
```
[the remark this answers is not in the document]
Comment (ricardoV94) [reply aesara/tensor/basic_opt.py:3529]: It seems this rewrite would basically just return the (stacked) output of `SpecifyShape.infer_shape`? Isn't this already done automatically when a shape graph is created?

Would this make sense?

```python
@register_useless
@register_canonicalize
@local_optimizer([Shape])
def local_Shape_of_SpecifyShape(fgraph, node):
    """Replace ``specify_shape(x, s).shape`` with ``s``."""
    if not isinstance(node.op, Shape):
        return False

    specified_shape = node.inputs[0]

    if not isinstance(getattr(specified_shape.owner, "op", None), SpecifyShape):
        return False

    shape = SpecifyShape.infer_shape(
        fgraph, 
        specified_shape.owner, 
        (inp.shape for inp in specified_shape.owner.inputs),
    )

    return [stack(shape).astype(np.int64)]
```

[the remark this answers is not in the document]
Comment (brandonwillard) [reply aesara/tensor/basic_opt.py:3529]: No, that would effectively bypass`ShapeFeature`.

[the remark this answers is not in the document]
Comment (ricardoV94) [reply aesara/tensor/basic_opt.py]: Making use of `shape_i` now, let me know if it looks correct

**Review by brandonwillard — APPROVED** [review]
This looks great!  Let's merge it.
